Sprint 12, Add database

cepheil opened the pull request.

cepheil committed:
feat: доработать модели и сервисы, добавить схему БД и README.md

- Film: добавить жанры и рейтинги
- Добавить новые сущности:  FriendshipStatus, Genre, RatingMpa
- User: добавить статус для связи «дружба» между пользователями
- UserService: доработать логику добавления в друзья
- UserController добавить метод подтверждения дружбы confirmFriendship
- Создать схему базы данных с таблицами и связями: schema.sql,  data.sql
- README.md: добавить диаграмму и описание схемы БД для наглядности

cepheil committed:
feat: доработать тесты для нового функционала, обязательные поля рейтинг и жанры для фильмов:
 FilmControllerTest
 FilmValidationTest
 в модели Genre и RatingMpa добавлены конструкторы.

 статус дружбы.
 UserServiceTest

cepheil committed:
feat: доработать тесты для нового функционала, обязательные поля рейтинг и жанры для фильмов:
 FilmControllerTest
 FilmValidationTest
 в модели Genre и RatingMpa добавлены конструкторы.

 статус дружбы.
 UserServiceTest

cepheil committed:
реализация ТЗ12, с расширенной логикой.

дружба имеет 3 статуса:
addFriend(user1, user2)
user1 : PENDING
user2 : REQUEST_RECEIVED (в обратную сторону.)

Если встречный запрос уже был — подтверждается обе стороны (CONFIRMED).

getFriends() возвращает только друзей со статусом CONFIRMED.

При удалении удаляются обе стороны.

Все sql запросы ориентированы на postgresql БД. (а тестирование проводится на Н2)

cepheil committed:
Добавлена реализация работы с базой данных

- Добавлены DAO-слои для основных сущностей: Film, User, Genre, RatingMpa, Friendship, Like
- Реализованы интерфейсы и их JDBC-реализации: FilmRepository, JdbcFilmRepository, FriendshipRepository, JdbcFriendshipRepository, GenreRepository, JdbcGenreRepository, LikeRepository, JdbcLikeRepository, RatingMpaRepository, JdbcRatingMpaRepository, UserRepository, JdbcUserRepository
- Созданы базовый класс BaseRepository и мапперы для преобразования SQL-результатов в модели: FilmRowMapper, GenreRowMapper, RatingMpaRowMapper, UserRowMapper
- Написаны интеграционные тесты для проверки корректности работы с базой данных
- Настроена работа с H2 в режиме, максимально приближенном к реальному окружению

cepheil committed:
Checkstyle fix

cepheil committed:
Checkstyle fix

cepheil committed:
Checkstyle fix

cepheil committed:
Checkstyle fix

cepheil committed:
validateUser fix

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/dal/JdbcFilmRepository.java:
```diff
@@ -0,0 +1,236 @@
+public class JdbcFilmRepository extends BaseRepository<Film> implements FilmRepository {
+    private final JdbcGenreRepository genreRepository;
+    private final JdbcLikeRepository likeRepository;
+    private final JdbcRatingMpaRepository ratingRepository;
```
private final JdbcGenreRepository genreRepository;
    private final JdbcLikeRepository likeRepository;
    private final JdbcRatingMpaRepository ratingRepository;

Вообще говоря, нет, это нехорошо. Репозитории (DAO) ничего друг про друга знать не должны. Репозиторий выполняет  crud команды для своей таблицы (набора данных) и,  если надо,  возвращает результаты в сервис. 
Если же репозиторий вызывает другие репозитории, то есть фактически зависит от них, то это серьёзный повод пересмотреть архитектуру программы. 
Вся бизнес логика находится в сервисах. Сервис может обращаться к другим сервисам, а также получать данные из нескольких репозиториев. Репозиторий же настроен на свою таблицу. 
.--------------------
Вся работа по сборке и обработке данных должна происходить в сервисе. Потому что это логика программы, то есть бизнес-правила, а сервисы как раз и предназначены для бизнес-правил.
А в репозитариях лучше оставить только базовые операции с соответствующей таблицей (набором данных).

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/dal/JdbcFilmRepository.java:
```diff
@@ -0,0 +1,236 @@
+
+        // Добавляем новые лайки
+        for (Long userId : film.getLikes()) {
+            likeRepository.addLike(film.getId(), userId);
```
В этом месте, в цикле, всё время дёргается база данных. Было бы хорошо избавиться от этих многочисленных запросов, которые сильно нагружают сервер. 
Хорошо было бы  сохранить данные одним запросом, например, с помощью конструкции:
jdbcTemplate.batchUpdate(............

Пакетная обработка в JDBC и HIBERNATE
https://habr.com/ru/articles/501756/

https://www.baeldung.com/jpa-hibernate-batch-insert-update

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/dal/JdbcFilmRepository.java:
```diff
@@ -0,0 +1,236 @@
+    @Override
+    public Collection<Film> getPopularFilms(int count) {
+        Collection<Film> films = findMany(GET_POPULAR_FILM_QUERY, count);
+        films.forEach(this::loadAdditionalData);
```
Здесь та же история-запросы в цикле. Нужно постараться избавиться..Сократить количество запросов.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/dal/JdbcFilmRepository.java:
```diff
@@ -0,0 +1,236 @@
+    @Override
+    public Collection<Film> findAll() {
+        Collection<Film> films = findMany(FIND_ALL_QUERY);
+        films.forEach(this::loadAdditionalData);
```
и здесь запросы к БД в цикле.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/service/EntityValidator.java:
```diff
@@ -0,0 +1,205 @@
+
+        if (film.getGenres() != null) {
+            for (Genre genre : film.getGenres()) {
+                validateGenreExists(genre.getId());
```
и здесь запросы к БД в цикле.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/service/EntityValidator.java:
```diff
@@ -0,0 +1,205 @@
+    }
+
+//    public void validateFilmUniqueness(Film film) {
+//        // Проверяем, существует ли фильм с таким же названием и датой релиза
```
Если код больше не нужен, его лучше удалить.

avfyodorov reviewed (commented):
Добрый день, Сергей! 

В целом хорошо, правда, есть ряд уточнений. 
Основная проблема - это нужно постараться избавиться от запросов к базе данных  из циклов.

cepheil committed:
Рефакторинг обработки данных фильмов и жанров:
 - Обновлены репозитории (Film, Genre, Like)
 - Скорректирована логика модели Film и сервисов
 - Рефакторинг EntityValidator

cepheil committed:
Checkstyle fix

avfyodorov reviewed (commented):
Добрый день, Сергей!

Замечаний нет.
Работа принята.
